File: kthSmallest.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function (root, k) {
  const t = [];
  const inOrder = (node) => {
    if (!node) return;
    inOrder(node.left);
    t.push(node.val);
    inOrder(node.right);
  };
  inOrder(root);
  return t[k - 1];
};

var kthSmallest = function (root, k) {
  let i = 0;
  let n = null;
  const inOrder = (node) => {
    if (node.left != null) inOrder(node.left);
    i += 1;
    if (i === k) {
      n = node.val;
      return;
    }
    if (node.right != null) inOrder(node.right);
  };
  inOrder(root);
  return n;
};
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}
const root = new TreeNode(
  5,
  new TreeNode(3, new TreeNode(2, new TreeNode(1)), new TreeNode(4)),
  new TreeNode(6)
);
kthSmallest(root, 4);

File: leetcode/kthSmallest.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
# class Solution:
#     def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
#         return kth(root, k)


def kth(root, k):
    res = {'i': 0, 'val': None}

    def in_order(node):
        if res['val']:
            return
        if node.left:
            in_order(node.left)
        res['i'] += 1
        if res['i'] == k:
            res['val'] = node.val
            return
        if node.right:
            in_order(node.right)

    in_order(root)
    return res['val']


def kth_array(root, k):
    res = []

    def in_order(node):
        if node:
            in_order(node.left)
            res.append(node.val)
            in_order(node.right)
    in_order(root)
    return res[k-1]


root = TreeNode(3, TreeNode(1, None, TreeNode(2)), TreeNode(4))
kth(root, 1)
